File: backend/app/routers/integrations.py
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, File, UploadFile
from starlette.concurrency import run_in_threadpool

from ..dependencies import Services, get_services
from ..schemas import (
    ScrivenerItemResponse,
    ProjectSyncsResponse,
    SyncBatchResponse,
    SyncConfigure,
    SyncDetachAllResponse,
    SyncRunResponse,
    SyncSummaryResponse,
    WordCountResponse,
    WordImportResponse,
)


router = APIRouter(tags=['integrations'])


@router.get('/projects/{project_id}/sync', response_model=SyncSummaryResponse)
def get_sync(
        project_id: str,
        services: Annotated[Services, Depends(get_services)],
        stage_id: str | None = None,
):
    return services.integrations.get_sync(project_id, stage_id=stage_id)


@router.get('/projects/{project_id}/sync/all', response_model=ProjectSyncsResponse)
def get_project_syncs(
        project_id: str,
        services: Annotated[Services, Depends(get_services)],
):
    return services.integrations.get_project_syncs(project_id)


@router.put('/projects/{project_id}/sync', response_model=SyncSummaryResponse)
def configure_sync(
        project_id: str,
        payload: SyncConfigure,
        services: Annotated[Services, Depends(get_services)],
):
    return services.integrations.configure_sync(
        project_id,
        sync_type=payload.type,
        path=payload.path,
        stage_id=payload.stage_id,
        item_id=payload.item_id,
    )


@router.delete('/projects/{project_id}/sync', response_model=SyncSummaryResponse)
def remove_sync(
        project_id: str,
        services: Annotated[Services, Depends(get_services)],
        stage_id: str | None = None,
):
    return services.integrations.remove_sync(project_id, stage_id=stage_id)


@router.delete(
    '/projects/{project_id}/sync/all',
    response_model=SyncDetachAllResponse,
)
def remove_all_syncs(
        project_id: str,
        services: Annotated[Services, Depends(get_services)],
):
    return services.integrations.remove_all_syncs(project_id)


@router.post('/projects/{project_id}/sync/run-all', response_model=SyncBatchResponse)
def run_project_syncs(
        project_id: str,
        services: Annotated[Services, Depends(get_services)],
):
    return services.integrations.run_project_syncs(project_id)


@router.post('/projects/{project_id}/sync/run', response_model=SyncRunResponse)
def run_sync(
        project_id: str,
        services: Annotated[Services, Depends(get_services)],
        stage_id: str | None = None,
):
    return services.integrations.run_sync(project_id, stage_id=stage_id)


@router.post('/integrations/sync/run-all', response_model=SyncBatchResponse)
def run_all_sync(services: Annotated[Services, Depends(get_services)]):
    return services.integrations.sync_all_configured()


@router.get(
    '/integrations/scrivener/items',
    response_model=list[ScrivenerItemResponse],
)
def scrivener_items(
        path: str,
        services: Annotated[Services, Depends(get_services)],
):
    return services.integrations.inspect_scrivener(path)


@router.post('/integrations/word/count', response_model=WordCountResponse)
async def count_word_upload(
        services: Annotated[Services, Depends(get_services)],
        file: UploadFile = File(...),
):
    content = await file.read(100 * 1024 * 1024 + 1)
    symbols = await run_in_threadpool(
        services.integrations.count_uploaded_docx,
        content,
        file.filename or 'document.docx',
    )
    return {'symbols': symbols}


@router.post(
    '/projects/{project_id}/imports/word',
    response_model=WordImportResponse,
)
async def apply_word_upload(
        project_id: str,
        services: Annotated[Services, Depends(get_services)],
        file: UploadFile = File(...),
        stage_id: str | None = None,
):
    content = await file.read(100 * 1024 * 1024 + 1)
    return await run_in_threadpool(
        services.integrations.apply_uploaded_docx,
        project_id,
        content,
        file.filename or 'document.docx',
        stage_id=stage_id,
    )

File: frontend/src/api/integrations.ts
import { apiRequest } from './client'
import { currentPlatform } from '@/platform/runtime'
import type {
  ScrivenerItem,
  ProjectSyncs,
  SyncBatchResult,
  SyncConfigure,
  SyncRunResult,
  SyncSummary,
  WordCountResult,
  WordImportResult,
} from '@/types/integrations'

function projectPath(projectId: string): string {
  return `/api/projects/${encodeURIComponent(projectId)}`
}

function projectSyncPath(projectId: string): string {
  return `${projectPath(projectId)}/sync`
}

function stageQuery(stageId?: string | null): string {
  if (!stageId) return ''
  return `?${new URLSearchParams({ stage_id: stageId }).toString()}`
}

async function nativeInvoke<T>(command: string, args?: Record<string, unknown>): Promise<T> {
  const { invoke } = await import('@tauri-apps/api/core')
  return invoke<T>(command, args)
}

export const integrationsApi = {
  getSync(projectId: string, stageId?: string | null): Promise<SyncSummary> {
    if (currentPlatform() === 'tauri') {
      return nativeInvoke<SyncSummary>('get_document_sync', { projectId, stageId: stageId ?? null })
    }
    return apiRequest<SyncSummary>(`${projectSyncPath(projectId)}${stageQuery(stageId)}`)
  },

  getProjectSyncs(projectId: string): Promise<ProjectSyncs> {
    if (currentPlatform() === 'tauri') {
      return nativeInvoke<ProjectSyncs>('get_project_document_syncs', { projectId })
    }
    return apiRequest<ProjectSyncs>(`${projectSyncPath(projectId)}/all`)
  },

  configureSync(projectId: string, payload: SyncConfigure): Promise<SyncSummary> {
    if (currentPlatform() === 'tauri') {
      return nativeInvoke<SyncSummary>('configure_document_sync', { projectId, ...payload })
    }
    return apiRequest<SyncSummary>(projectSyncPath(projectId), {
      method: 'PUT',
      body: payload,
    })
  },

  removeSync(projectId: string, stageId?: string | null): Promise<SyncSummary> {
    if (currentPlatform() === 'tauri') {
      return nativeInvoke<SyncSummary>('remove_document_sync', { projectId, stageId: stageId ?? null })
    }
    return apiRequest<SyncSummary>(`${projectSyncPath(projectId)}${stageQuery(stageId)}`, {
      method: 'DELETE',
    })
  },

  runSync(projectId: string, stageId?: string | null): Promise<SyncRunResult> {
    if (currentPlatform() === 'tauri') {
      return nativeInvoke<SyncRunResult>('run_document_sync', { projectId, stageId: stageId ?? null })
    }
    return apiRequest<SyncRunResult>(
      `${projectSyncPath(projectId)}/run${stageQuery(stageId)}`,
      { method: 'POST' },
    )
  },

  runProjectSyncs(projectId: string): Promise<SyncBatchResult> {
    if (currentPlatform() === 'tauri') {
      return nativeInvoke<SyncBatchResult>('run_project_document_syncs', { projectId })
    }
    return apiRequest<SyncBatchResult>(`${projectSyncPath(projectId)}/run-all`, {
      method: 'POST',
    })
  },

  runAllSync(): Promise<SyncBatchResult> {
    if (currentPlatform() === 'tauri') {
      return nativeInvoke<SyncBatchResult>('run_all_document_sync')
    }
    return apiRequest<SyncBatchResult>('/api/integrations/sync/run-all', {
      method: 'POST',
    })
  },

  inspectScrivener(path: string): Promise<ScrivenerItem[]> {
    if (currentPlatform() === 'tauri') {
      return nativeInvoke<ScrivenerItem[]>('inspect_scrivener', { path })
    }
    const params = new URLSearchParams({ path })
    return apiRequest<ScrivenerItem[]>(`/api/integrations/scrivener/items?${params.toString()}`)
  },

  countWord(file: File): Promise<WordCountResult> {
    if (currentPlatform() === 'tauri') {
      return file.arrayBuffer().then((buffer) => nativeInvoke<number>('count_word_document', {
        bytes: Array.from(new Uint8Array(buffer)),
        filename: file.name,
      }).then((symbols) => ({ symbols })))
    }
    const body = new FormData()
    body.set('file', file, file.name)
    return apiRequest<WordCountResult>('/api/integrations/word/count', {
      method: 'POST',
      rawBody: body,
    })
  },

  importWord(
    projectId: string,
    file: File,
    stageId?: string | null,
  ): Promise<WordImportResult> {
    if (currentPlatform() === 'tauri') {
      return file.arrayBuffer().then((buffer) => nativeInvoke<WordImportResult>('import_word_document', {
        projectId,
        stageId: stageId ?? null,
        bytes: Array.from(new Uint8Array(buffer)),
        filename: file.name,
      }))
    }
    const body = new FormData()
    body.set('file', file, file.name)
    return apiRequest<WordImportResult>(
      `${projectPath(projectId)}/imports/word${stageQuery(stageId)}`,
      { method: 'POST', rawBody: body },
    )
  },
}
